IdentityCore/src/util/MSIDXpcSingleSignOnProvider.m

```diff
@@ -0,0 +1,330 @@
+    status = SecCodeCopySigningInformation(selfCode, kSecCSSigningInformation, &cfDic);
+    
+    if (!cfDic)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, [NSString stringWithFormat:@"Failed to copy signing dictionary, status: %ld", (long)status], nil, nil);
+        CFRelease(selfCode);
+        return nil;
+    }
+    
+    NSDictionary *signingDic = CFBridgingRelease(cfDic);
+    CFRelease(selfCode);
+    
+    return [self devSigningIdentityFromSigningDictionary:signingDic];
+}
+
+- (NSString *)devSigningIdentityFromSigningDictionary:(NSDictionary *)signingDic
+{
+    NSArray *signingCertificates = signingDic[@"certificates"];
+    
+    if (!signingCertificates
+        || ![signingCertificates isKindOfClass:[NSArray class]]
+        || ![signingCertificates count])
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, @"No certificates present in the signing dictionary", nil, nil);
+        return nil;
+    }
+    
+    SecCertificateRef leafCert = (__bridge SecCertificateRef)(signingCertificates[0]);
+    CFStringRef certCommonName = NULL;
+    OSStatus status = SecCertificateCopyCommonName(leafCert, &certCommonName);
+    
+    if (!certCommonName)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, [NSString stringWithFormat:@"Error with code %ld, description %@", (long)status, @"Failed to copy certificate common name"], nil, nil);
+        return nil;
+    }
+    
+    return CFBridgingRelease(certCommonName);
+}
+
+- (void)forceRunOnBackgroundQueue:(BOOL)forceOnBackgroundQueue dispatchBlock:(void (^)(void))dispatchBlock
+{
+    if (forceOnBackgroundQueue && [NSThread isMainThread])
+    {
+        dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
+            dispatchBlock();
+        });
+    }
+    else
+    {
+        dispatchBlock();
+    }
+}
+
+- (void)getXpcService:(NSXPCListenerEndpointCompletionBlock)continueBlock
+{
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, @"[Entra broker] CLIENT - started establishing connection");
+    NSXPCConnection *connection = [[NSXPCConnection alloc] initWithMachServiceName:machServiceName options:0];
+    
+    NSString *codeSigningRequirement = [self codeSignRequirementForBundleId:brokerDispatcher devIdentity:[self signingIdentity]];
+    
+    connection.remoteObjectInterface = [NSXPCInterface interfaceWithProtocol:@protocol(MSIDXpcBrokerDispatcherProtocol)];
+    if (@available(macOS 13.0, *)) {
+        [connection setCodeSigningRequirement:codeSigningRequirement];
+    } else {
+        // Intentionally left empty because the entire XPC flow will only be available on macOS 13 and above and gaurded through canPerformRequest
+    }
+    
+    [connection resume];
+    [connection setInterruptionHandler:^{
+        NSError *xpcUnexpectedError = MSIDCreateError(MSIDErrorDomain, MSIDErrorSSOExtensionUnexpectedError, @"[Entra broker] CLIENT -- dispatcher connection is interrupted", nil, nil, nil, nil, nil, YES);
```

Comment:
Do we want a different error such that we can separate SSO extension unexpected error v.s. XPC unexpected error? :)

Reply:
Yes, I have a different item to add a new error types